design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,189 @@
+# KLIP 15 - KQSLDB Client and new API
+
+**Author**: Tim Fox (purplefox) | 
+**Release Target**: ? | 
+**Status**: _In Discussion_ | 
+**Discussion**:
+
+Please read the initial discussion on the ksqlDB developer group first:
+
+https://groups.google.com/forum/#!topic/ksql-dev/yxcRlsOsNmo
+
+And there is a partially working prototype for this work described here:
+
+https://groups.google.com/forum/#!topic/ksql-dev/5mLKvtZFs4Y
+
+ For ksqlDB to be a successful project we need to provide an awesome out of the box developer experience
+ and it should be super easy and fun to write powerful event streaming applications.
+ Currently this is somewhat diffificult to do, and it requires the use of multiple clients and multiple moving parts in order to build
+ a simple event sourcing / CQRS style application.
+ We should provide a new ksqlDB client that provides that currently missing single interaction point with the event streaming
+ platform so that we can improve the application development process.
+           
+## Motivation and background
+
+In order to increase adoption of ksqlDB we need the developer experience to be as slick as possible. Currently, in order to write
+a simple event streaming application with ksqlDB, up to 4 clients may be needed: 
+
+1. An HTTP client to access the current HTTP/REST API
+2. A WebSocket client to access the streaming query WebSockets endpoint
+3. A JDBC client to access data in JDBC tables that are synced with KSQL via connect, as the current support for
+pull queries does not make the data otherwise easily accessible.
+4. A Kafka client for producing and consuming messages from topics.
+
+This is a lot for an application to handle and creates a steep learning curve for application developers.
+
+Moreover, our current HTTP API is lacking, especially in the area of streaming query results which makes it difficult to write an application
+that uses this API.
+
+This KLIP proposes that we:
+
+* Create a ksqlDB client (initially in Java and JavaScript) as the primary interaction point to the event streaming platform
```

Comment:
To further the KLIP discussion from the JavaScript perspective, I opened a PR against @purplefox's prototype branch: https://github.com/purplefox/ksql/pull/12.

From the PR description: 

>Provides a JS (TypeScript) implementation of the wire protocol and client semantics proposed in KLIP-15. The client works in Node.js and browser runtimes.

> The goals of this exercise are to:
> 1. Demonstrate a non-Java impl of KLIP-15's novel wire protocol.
> 2. Consider the ideal client API for consuming streaming queries in JavaScript.
> 3. Take the perspective of a front-end developer building a web app powered by ksqlDB.

> The included example ports the KLIP's `ksql-real-app` to a single-page web app:
> ![shopping-cart](https://user-images.githubusercontent.com/141130/70751751-1f902500-1cff-11ea-9fde-3ede3ff23d18.gif)

> See the [README](https://github.com/purplefox/ksql/blob/f38fe8c1a97c1a29c526a0fe5cad58a68d70e316/ksql-js-client/README.md) for information on how to run the example.

I will follow up in this thread with more thoughts on the outlined goals above.

Reply:
> 1. Demonstrate a non-Java impl of KLIP-15's novel wire protocol.

I found it pretty straightforward to [port the frame types and their binary encoding](https://github.com/purplefox/ksql/blob/320e5aeb73db1d4f9a4f4036ab61aa39b405dd14/ksql-js-client/packages/client/src/protocol.ts) based on the Java implementation. I wrote [some tests](https://github.com/purplefox/ksql/blob/320e5aeb73db1d4f9a4f4036ab61aa39b405dd14/ksql-js-client/packages/client/test/protocol.ts) to confirm my understanding. And for sure there are many opportunities for optimization in my sample port.

It was a bit more challenging to build a mental model for the frame sequencing based only on the Java reference impl, particularly (and naturally so) for [handling `streamQuery` semantics](https://github.com/purplefox/ksql/blob/320e5aeb73db1d4f9a4f4036ab61aa39b405dd14/ksql-js-client/packages/client/src/query-result.ts#L64-L120). Still, it was totally tractable, since the protocol is so simple.

Assuming we document the protocol specification as part of the KLIP implementation, I think folks would have a relatively easy time porting the suggested protocol to other languages.

Reply:
> 2. Consider the ideal client API for consuming streaming queries in JavaScript.

The client API for flowing push query results through a JS application is a principal opportunity to tell the reactive streaming story. On the flip side, if the developer experience for consuming pushes sucks, we're not doing ksqlDB justice. Same for Java.

> Should streamQuery results return the JS implementation of `concurrent.Flow`?

While there is an argument for emulating the Java client usage patterns, this seems overly prescriptive. Flowable has little traction in the JS world. In the worst case this approach creates a perception that the JS client is second-class.

> What about [RxJS Observables](http://reactivex.io/documentation/observable.html)?

RxJS is de facto-ish for reactive patterns. Folks already using this library would be delighted. But if you are not acquainted with the Rx way, the breadth of concepts it prescribes knowing could be intimidating.

> What is the case for [`AsyncIterable`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of)?

It's part of the ES 2018 standard. Its cross section is way smaller than the RxJS option.

> Do we need to offer `poll()` in addition?

Probably. Familiarity is important for folks and AsyncIterable is new. Returning a simple event emitter might be the best bet.

> What about client API methods other than `streamQuery` (with pushes)?

For other request patterns, there's not as much to consider. In general, it's natural to model the async work as a Promise in the same cases for which we return CompletableFuture in Java.

Reply:
> 3. Take the perspective of a front-end developer building a web app powered by ksqlDB.

For me this was the fun part. FWIW, I did it after the bulk of the client implementation, and it felt really good to make stuff tangible. The [React port](https://github.com/purplefox/ksql/tree/320e5aeb73db1d4f9a4f4036ab61aa39b405dd14/ksql-js-client/packages/shopping-cart-example/src) of the `ksql-real-app` example is very little code.

Here's a way to deliver push queries with React hooks:
```js
import React, { useState, useEffect } from 'react';

// where `db` is the ksqlDB client connection
 const usePushQuery = (db, query) => {
  const [results, setResults] = useState({ rows: [], error: null });

  useEffect(() => {
    (async () => {
      const rows = [];
      try {
        const pushResults = await db.streamQuery(query);
        for await (const row of pushResults) {
          rows.push(row);
          setResults({ rows });
        }
      } catch (error) {
        setResults({ error });
      }
    })();
  }, [query]);

  return results;
};

// usage in component fn
  const { rows, error } = usePushQuery(db, `SELECT * FROM ORDER_REPORT EMIT CHANGES`);
```

ksqlDB + JS can be as delightful as it is powerful.

Reply:
This is awesome :)